packages/cli/src/verify.ts

```diff
@@ -0,0 +1,80 @@
+import { forge, getRpcUrl } from "@latticexyz/common/foundry";
+import { Hex, createWalletClient, getCreate2Address, http } from "viem";
+import { salt } from "./deploy/common";
+import { ensureDeployer } from "./deploy/ensureDeployer";
+import { privateKeyToAccount } from "viem/accounts";
+import { MUDError } from "@latticexyz/common/errors";
+import { getWorldFactoryContracts } from "./deploy/ensureWorldFactory";
+import { Contract } from "./deploy/ensureContract";
+
+type VerifyOptions = {
+  foundryProfile?: string;
+  systems: Contract[];
+  modules: Contract[];
+  worldAddress?: Hex;
+};
+
+async function verifyContract(
+  foundryProfile: string | undefined,
+  deployerAddress: Hex,
+  contract: Contract,
+  rpc: string,
+  cwd?: string,
+) {
+  if (!contract.label) {
+    throw new MUDError("Need a label");
+  }
+
+  const system = getCreate2Address({ from: deployerAddress, bytecode: contract.bytecode, salt });
+
+  await forge(["verify-contract", system, contract.label, "--verifier", "sourcify", "--rpc-url", rpc], {
+    profile: foundryProfile,
+    cwd,
+  });
+}
+
+export async function verify({
+  foundryProfile = process.env.FOUNDRY_PROFILE,
+  systems,
+  modules,
+  worldAddress,
+}: VerifyOptions): Promise<void> {
+  const privateKey = process.env.PRIVATE_KEY as Hex;
+  if (!privateKey) {
+    throw new MUDError(
+      `Missing PRIVATE_KEY environment variable.
+Run 'echo "PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80" > .env'
+in your contracts directory to use the default anvil private key.`,
+    );
+  }
+
+  const rpc = await getRpcUrl(foundryProfile);
+
+  const client = createWalletClient({
+    transport: http(rpc),
+    account: privateKeyToAccount(privateKey),
+  });
+
+  const deployerAddress = await ensureDeployer(client);
+
+  await Promise.all(systems.map((contract) => verifyContract(foundryProfile, deployerAddress, contract, rpc)));
+
+  await Promise.all(
+    getWorldFactoryContracts(deployerAddress).map((contract) =>
+      verifyContract(foundryProfile, deployerAddress, contract, rpc, "node_modules/@latticexyz/world"),
+    ),
+  );
+
+  await Promise.all(
+    modules.map((contract) =>
+      verifyContract(foundryProfile, deployerAddress, contract, rpc, "node_modules/@latticexyz/world-modules"),
+    ),
+  );
```

Comment:
can these be parallelized with the above?

```
await Promise.all([
  ...getWorldFactoryContracts(...).map(...),
  modules.map(...),
]);
```

Reply:
and with the parallelizing, anything to be aware of as far as sourcify rate limits?

if so, we can wrap this in `p-queue` to ensure some level of concurrency

Reply:
With this method I do get odd behaviour from sourcify, lemme try a p-queue

Reply:
no worries if that doesn't work, there could be an order of operations thing by sourcify

it's the same reason we deploy certain chunks of contracts before others, to ensure the previous ones are there before the new ones because they depend on em and will fail otherwise